internal/eval/store_eval.go

```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
 
 	return evalSET(newArgs, store)
 }
+
+// evalHEXISTS returns if field is an existing field in the hash stored at key.
+//
+// This command returns 0, if the specified field doesn't exist in the key and 1 if it exists.
+//
+// If key doesn't exist, it returns 0.
+//
+// Usage: HEXISTS key field
+func evalHEXISTS(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) != 2 {
+		return &EvalResponse{
+			Result: nil,
+			Error:  diceerrors.ErrWrongArgumentCount("HEXISTS"),
+		}
+	}
+
+	key := args[0]
+	hmKey := args[1]
+	obj := store.Get(key)
+
+	var hashMap HashMap
+
+	if obj == nil {
+		return &EvalResponse{
+			Result: clientio.Encode(0, false),
+			Error:  nil,
+		}
+	}
+	if err := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeHashMap, object.ObjEncodingHashMap); err != nil {
+		return &EvalResponse{
+			Error:  diceerrors.ErrGeneral(diceerrors.WrongTypeErr),
+			Result: nil,
+		}
+	}
+
+	hashMap = obj.Value.(HashMap)
+
+	_, ok := hashMap.Get(hmKey)
+	if ok {
+		return &EvalResponse{
+			Result: clientio.Encode(1, false),
+			Error:  nil,
+		}
+	}
+	// Return 0, if specified field doesn't exist in the HashMap.
+	return &EvalResponse{
+		Result: clientio.Encode(0, false),
+		Error:  nil,
+	}
+}
+
+// evalHKEYS is used to retrieve all the keys(or field names) within a hash.
+//
+// This command returns empty array, if the specified key doesn't exist.
+//
+// Complexity is O(n) where n is the size of the hash.
+//
+// Usage: HKEYS key
+func evalHKEYS(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) != 1 {
+		return &EvalResponse{
+			Result: nil,
+			Error:  diceerrors.ErrWrongArgumentCount("HKEYS"),
+		}
+	}
+
+	key := args[0]
+	obj := store.Get(key)
+
+	var hashMap HashMap
+	var result []string
+
+	if obj != nil {
+		if err := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeHashMap, object.ObjEncodingHashMap); err != nil {
+			return &EvalResponse{
+				Error:  diceerrors.ErrGeneral(diceerrors.WrongTypeErr),
+				Result: nil,
+			}
+		}
+		hashMap = obj.Value.(HashMap)
+	} else {
+		return &EvalResponse{
+			Result: clientio.Encode([]interface{}{}, false),
+			Error:  nil,
+		}
+	}
+
+	for hmKey := range hashMap {
+		result = append(result, hmKey)
+	}
+
+	return &EvalResponse{
+		Result: clientio.Encode(result, false),
+		Error:  nil,
+	}
+}
+
+// evalHKEYS returns all the values in the hash stored at key.
+func evalHVALS(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) != 1 {
+		return &EvalResponse{Error: diceerrors.ErrWrongArgumentCount("HVALS"), Result: nil}
+	}
+
+	key := args[0]
+	obj := store.Get(key)
+
+	if obj == nil {
+		// Return an empty array for non-existent keys
+		return &EvalResponse{
+			Result: clientio.Encode([]string{}, false),
```

Comment:
Please send the response without encoding. Raw results that makes it protocol independent.